Add a test of minimal stacks, and make it pass

ctiller opened the pull request:
Optionalizes a bunch of filters and moves them to the ext/ tree

The new test serves as documentation for what's meant by a minimal stack

Built on #10423

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/10449)
<!-- Reviewable:end -->

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

markdroth commented:
Reviewed 58 of 78 files at r1, 7 of 9 files at r2, 6 of 8 files at r3, 12 of 12 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[BUILD, line 753 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2esc3gyju-DtbJBx0:-Kh2esc3gyju-DtbJBx1:b-9884il) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/BUILD#L753)):*
> ```Python
> 
> grpc_cc_library(
>     name = "grpc_deadline_filter",
> ```

Out of curiosity, in what build will deadline support be optional?

---

*[BUILD, line 783 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2f2GCV4GGKFUM1lLQ:-Kh2f2GCV4GGKFUM1lLR:b-wcmy01) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/BUILD#L783)):*
> ```Python
>     name = "grpc_http_filters",
>     hdrs = [
>         "src/core/ext/filters/http/compress/compress_filter.h",
> ```

While we're doing this, would it make sense to rename this to something like `message_compress_filter`, so that it's clearly disambiguated from the upcoming stream compression code?

---

*[build.yaml, line 176 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2fOkT_EzvLpBv8e-7:-Kh2fOkT_EzvLpBv8e-8:b-g8kycw) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/build.yaml#L176)):*
> ```YAML
>   - include/grpc/status.h
>   headers:
>   - src/core/ext/filters/deadline/deadline_filter.h
> ```

I think this should be removed, since it's been moved to its own target.

---

*[build.yaml, line 283 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2fawqTRLsQTrXo84T:-Kh2fawqTRLsQTrXo84U:brhumzt) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/build.yaml#L283)):*
> ```YAML
>   - src/core/lib/transport/transport_impl.h
>   src:
>   - src/core/ext/filters/deadline/deadline_filter.c
> ```

Same here.

---

*[include/grpc/impl/codegen/grpc_types.h, line 181 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2g-GgZfP-aCL6bhpd:-Kh2g-GgZfP-aCL6bhpe:bz8eyfs) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/include/grpc/impl/codegen/grpc_types.h#L181)):*
> /** Enable/disable support for deadline checking. Defaults to 1, unless
>     GRPC_ARG_MINIMAL_STACK is enabled, in which case it defaults to 0 */

Do we need to say something here like "...unless built without deadline support"?

---

*[src/core/ext/filters/client_channel/client_channel.c, line 1281 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2gflnxYn7B6w9P3Ng:-Kh2gflnxYn7B6w9P3Nh:b47d71y) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/client_channel/client_channel.c#L1281)):*
>     grpc_deadline_state_init(exec_ctx, elem, args->call_stack);
>     grpc_deadline_state_start(exec_ctx, elem, calld->deadline);

These two methods could probably be combined, since I don't think either of them are actually called from anywhere else anymore.

---

*[src/core/ext/filters/deadline/deadline_filter.c, line 356 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2h8ggLvIIavEpuVTt:-Kh2h8ghh8AWylCyX_F6:b-12xnrv) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/deadline/deadline_filter.c#L356)):*
> ```C
>   const grpc_arg* a =
>       grpc_channel_args_find(channel_args, GRPC_ARG_ENABLE_DEADLINE_CHECKS);
>   if (a != NULL && a->type == GRPC_ARG_INTEGER && a->value.integer != 0) {
> ```

Let's use `grpc_channel_arg_get_integer()` here.

---

*[src/core/ext/filters/deadline/deadline_filter.c, line 357 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2hP-PinNGqN_WG97N:-Kh2hP-PinNGqN_WG97O:bi6q47z) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/deadline/deadline_filter.c#L357)):*
> ```C
>       grpc_channel_args_find(channel_args, GRPC_ARG_ENABLE_DEADLINE_CHECKS);
>   if (a != NULL && a->type == GRPC_ARG_INTEGER && a->value.integer != 0) {
>     enable = true;
> ```

Shouldn't this be `&=`?  Otherwise, we're ignoring the value of the want_minimal_stack arg in favor of the enable_deadline_checks arg.

---

*[src/core/ext/filters/http/http_filters_plugin.c, line 54 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2jF0uFcMawGpCW2uO:-Kh2jF0uFcMawGpCW2uP:bpofuk) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/http/http_filters_plugin.c#L54)):*
> ```C
>     grpc_channel_stack_builder *builder) {
>   grpc_transport *t = grpc_channel_stack_builder_get_transport(builder);
>   return t && strstr(t->vtable->name, "http");
> ```

`t != NULL`

---

*[src/core/ext/filters/http/http_filters_plugin.c, line 68 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2jkSZYlJ3JtB4fHqI:-Kh2jkS_4324mYjseWj7:b-faun3r) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/http/http_filters_plugin.c#L68)):*
> ```C
>       grpc_channel_args_find(channel_args, filtarg->control_channel_arg);
>   if (ctlarg != NULL) {
>     enable = !(ctlarg->type == GRPC_ARG_INTEGER && ctlarg->value.integer == 0);
> ```

Shouldn't this be `&=`, so that we don't ignore the want_minimal_stack value from above?

---

*[src/core/ext/filters/http/http_filters_plugin.c, line 68 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2ju42QtXQfYtDrSjN:-Kh2ju42QtXQfYtDrSjO:b-12xnrv) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/http/http_filters_plugin.c#L68)):*
> ```C
>       grpc_channel_args_find(channel_args, filtarg->control_channel_arg);
>   if (ctlarg != NULL) {
>     enable = !(ctlarg->type == GRPC_ARG_INTEGER && ctlarg->value.integer == 0);
> ```

Let's use `grpc_channel_arg_get_integer()` here.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 94 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2lSAKDrkQTB0bALMf:-Kh2lSAKDrkQTB0bALMg:bw3zohm) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L94)):*
>   int max_send_size;
>   int max_recv_size;

Perhaps it would make sense to just replace these two fields with a `message_size_limits` struct.  That way, we could initialize both fields at once by just copying the struct.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 107 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2nY4QydhiuXk8i9U8:-Kh2nY4RpxF07z9RpW8J:b6kya13) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L107)):*
>   int max_send_size;
>   int max_recv_size;

Same here.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 214 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2k_nGslOPCfu6mlLx:-Kh2k_nGslOPCfu6mlLy:bfdia9z) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L214)):*
> ```C
>                         int without_minimal_stack) {
>   if (grpc_channel_args_want_minimal_stack(args)) {
>     return INT_MAX;
> ```

The existing code treats -1 as unlimited, so let's do the same thing here.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 219 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2l4Fxcp_7evRZcVB0:-Kh2l4Fxcp_7evRZcVB1:bczxbvo) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L219)):*
> typedef struct {
>   int max_recv_size;
>   int max_send_size;
> } channel_limits;

This looks exactly the same as the `message_size_limits` struct defined above.  Might as well re-use it.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 234 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2ln6Ik0YrVjvp8jAz:-Kh2ln6JIHR0xu0lThIv:ba8l3cr) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L234)):*
> ```C
>         0) {
>       const grpc_integer_options options = {
>           default_size(channel_args, GRPC_DEFAULT_MAX_SEND_MESSAGE_LENGTH), 0,
> ```

Could use `lim.max_send_size` here instead of calling `default_size()` again.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 242 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2lsJDnN9iUAAFzPCp:-Kh2lsJDnN9iUAAFzPCq:bvseiah) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L242)):*
> ```C
>                GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH) == 0) {
>       const grpc_integer_options options = {
>           default_size(channel_args, GRPC_DEFAULT_MAX_RECV_MESSAGE_LENGTH), 0,
> ```

Could use `lim.max_recv_size` here instead of calling `default_size()` again.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 305 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2oZ79fATNwXmd6wGz:-Kh2oZ79fATNwXmd6wH-:b-vst7n6) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L305)):*
>   channel_limits lim = get_channel_limits(channel_args);
>   if (lim.max_send_size != INT_MAX || lim.max_recv_size != INT_MAX) {
>     enable = true;
>   }
>   const grpc_arg* a =
>       grpc_channel_args_find(channel_args, GRPC_ARG_SERVICE_CONFIG);
>   if (a != NULL) {
>     enable = true;
>   }

Note that `GRPC_ARG_SERVICE_CONFIG` is only set when injecting a service config via the client API.  It will very rarely be set, but even when it's not, the resolver can still return a service config.  So I think we should unconditionally enable this filter, unless want_minimal_stack is set.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/10449)*
<!-- Sent from Reviewable.io -->

ctiller commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[BUILD, line 753 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2esc3gyju-DtbJBx0:-Kh2z--vMojCYFgOfkTi:b-tw4q02) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/BUILD#L753)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Out of curiosity, in what build will deadline support be optional?
</blockquote></details>

I'm not sure yet... but being a highly desirable but not strictly required thing still places it in ext/.

---

*[BUILD, line 783 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2f2GCV4GGKFUM1lLQ:-Kh3-b2cz7xLL8yE1OR_:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/BUILD#L783)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

While we're doing this, would it make sense to rename this to something like `message_compress_filter`, so that it's clearly disambiguated from the upcoming stream compression code?
</blockquote></details>

Done.

---

*[build.yaml, line 176 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2fOkT_EzvLpBv8e-7:-Kh3-kjfXKcFF-wTPK2V:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/build.yaml#L176)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I think this should be removed, since it's been moved to its own target.
</blockquote></details>

Done.

---

*[build.yaml, line 283 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2fawqTRLsQTrXo84T:-Kh3-lGkvY1ZRDIorzvo:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/build.yaml#L283)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Same here.
</blockquote></details>

Done.

---

*[include/grpc/impl/codegen/grpc_types.h, line 181 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2g-GgZfP-aCL6bhpd:-Kh3-oOKDSQ7DxiS3kga:b-90gc3p) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/include/grpc/impl/codegen/grpc_types.h#L181)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

> /** Enable/disable support for deadline checking. Defaults to 1, unless
>     GRPC_ARG_MINIMAL_STACK is enabled, in which case it defaults to 0 */

Do we need to say something here like "...unless built without deadline support"?
</blockquote></details>

For the moment no... I have no plans of actually offering a build without deadline support - but I do have a need to be able to call it out.

---

*[src/core/ext/filters/client_channel/client_channel.c, line 1281 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2gflnxYn7B6w9P3Ng:-Kh30xue2oDEBjSzCoH6:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/client_channel/client_channel.c#L1281)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

>     grpc_deadline_state_init(exec_ctx, elem, args->call_stack);
>     grpc_deadline_state_start(exec_ctx, elem, calld->deadline);

These two methods could probably be combined, since I don't think either of them are actually called from anywhere else anymore.
</blockquote></details>

Done.

---

*[src/core/ext/filters/deadline/deadline_filter.c, line 356 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2h8ggLvIIavEpuVTt:-Kh37a9I94JvUDWzk4aB:ba01ksr) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/deadline/deadline_filter.c#L356)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Let's use `grpc_channel_arg_get_integer()` here.
</blockquote></details>

Added a get_bool, used it.

---

*[src/core/ext/filters/deadline/deadline_filter.c, line 357 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2hP-PinNGqN_WG97N:-Kh31-v107hcDZFZXoff:b-qfjlxo) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/deadline/deadline_filter.c#L357)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Shouldn't this be `&=`?  Otherwise, we're ignoring the value of the want_minimal_stack arg in favor of the enable_deadline_checks arg.
</blockquote></details>

Minimal stack changes defaults, but still allows things to be turned on over it. So I can say minimal_stack + deadline, or I can say normal_stack - deadline.

Folks that want flat out performance and a few features get pushed towards minimal stack, and get to track what they need enabled.

98% of users use the default stack, and maybe turn off a few things they don't need if they're causing problems.

---

*[src/core/ext/filters/http/http_filters_plugin.c, line 54 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2jF0uFcMawGpCW2uO:-Kh37cWq8vJuyXTEUZGk:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/http/http_filters_plugin.c#L54)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

`t != NULL`
</blockquote></details>

Done.

---

*[src/core/ext/filters/http/http_filters_plugin.c, line 68 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2jkSZYlJ3JtB4fHqI:-Kh31S6m6VO-1MQpi-DC:b-34vmyd) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/http/http_filters_plugin.c#L68)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Shouldn't this be `&=`, so that we don't ignore the want_minimal_stack value from above?
</blockquote></details>

See above

---

*[src/core/ext/filters/http/http_filters_plugin.c, line 68 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2ju42QtXQfYtDrSjN:-Kh35SqZpSZ6ilFXIp0D:b-gybswo) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/http/http_filters_plugin.c#L68)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Let's use `grpc_channel_arg_get_integer()` here.
</blockquote></details>

Added a get_bool, used it all over the place

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 94 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2lSAKDrkQTB0bALMf:-Kh36MLN4KjfI1SpGsU4:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L94)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

>   int max_send_size;
>   int max_recv_size;

Perhaps it would make sense to just replace these two fields with a `message_size_limits` struct.  That way, we could initialize both fields at once by just copying the struct.
</blockquote></details>

Done.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 107 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2nY4QydhiuXk8i9U8:-Kh36Mc4WXuwFpJa-4LE:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L107)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

>   int max_send_size;
>   int max_recv_size;

Same here.
</blockquote></details>

Done.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 214 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2k_nGslOPCfu6mlLx:-Kh3749Im8Nkx7ygs4l6:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L214)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

The existing code treats -1 as unlimited, so let's do the same thing here.
</blockquote></details>

Done.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 219 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2l4Fxcp_7evRZcVB0:-Kh373ng-_1t-kIEq-lJ:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L219)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

> typedef struct {
>   int max_recv_size;
>   int max_send_size;
> } channel_limits;

This looks exactly the same as the `message_size_limits` struct defined above.  Might as well re-use it.
</blockquote></details>

Done.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 234 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2ln6Ik0YrVjvp8jAz:-Kh37GvCu_Z-oqDM9Z62:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L234)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Could use `lim.max_send_size` here instead of calling `default_size()` again.
</blockquote></details>

Done.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 242 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2lsJDnN9iUAAFzPCp:-Kh37GHjTwwJyGURFQi-:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L242)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Could use `lim.max_recv_size` here instead of calling `default_size()` again.
</blockquote></details>

Done.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 305 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2oZ79fATNwXmd6wGz:-Kh37IG2KCZXV11S3jhG:b6ll33z) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L305)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

>   channel_limits lim = get_channel_limits(channel_args);
>   if (lim.max_send_size != INT_MAX || lim.max_recv_size != INT_MAX) {
>     enable = true;
>   }
>   const grpc_arg* a =
>       grpc_channel_args_find(channel_args, GRPC_ARG_SERVICE_CONFIG);
>   if (a != NULL) {
>     enable = true;
>   }

Note that `GRPC_ARG_SERVICE_CONFIG` is only set when injecting a service config via the client API.  It will very rarely be set, but even when it's not, the resolver can still return a service config.  So I think we should unconditionally enable this filter, unless want_minimal_stack is set.
</blockquote></details>

How does the resolver notify this code of a service config when it's enabled? We should check that here.

There needs to be a very high bar for unconditionally enabling anything, and if we can detect something can be disabled, we should do so.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/10449)*
<!-- Sent from Reviewable.io -->

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                       allocs_per_iteration    atm_add_per_iteration    atm_cas_per_iteration      cpu_time  locks_per_iteration      real_time  writes_per_iteration
--------------------------------------------------------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_HpackEncoderEncodeHeader<SingleInternedBinaryElem<1>>/0/16k                                                                                 +9.00                               +9.00
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

markdroth commented:
Reviewed 5 of 80 files at r1, 38 of 40 files at r6, 1 of 1 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/core/ext/filters/message_size/message_size_filter.c, line 305 at r5](https://reviewable.io:443/reviews/grpc/grpc/10449#-Kh2oZ79fATNwXmd6wGz:-Kh4HeBnoQsuL6JSPlka:b-pe10fa) ([raw file](https://github.com/grpc/grpc/blob/ff558bac5c7be1571db720f23974d70b75385415/src/core/ext/filters/message_size/message_size_filter.c#L305)):*
<details><summary><i>Previously, ctiller (Craig Tiller) wrote…</i></summary><blockquote>

How does the resolver notify this code of a service config when it's enabled? We should check that here.

There needs to be a very high bar for unconditionally enabling anything, and if we can detect something can be disabled, we should do so.
</blockquote></details>

Hmm... There may not be an issue here after all.

I was thinking that this function gets run at channel stack construction time, which is long before the resolver runs.  However, that's only true of the parent channel stack; the subchannel stack actually does get run based on the args returned by the resolver.  And the resolver will in fact set `GRPC_ARG_SERVICE_CONFIG` to signal to modules like this one (which lives in the subchannel stack) that there's a service config.  So I think this is fine.  Sorry for the confusion.

Nothing to see here, please move along... :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/10449)*
<!-- Sent from Reviewable.io -->

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                   allocs_per_iteration    atm_add_per_iteration    atm_cas_per_iteration      cpu_time  locks_per_iteration      real_time  writes_per_iteration
----------------------------------------------------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_IsolatedFilter<ServerDeadlineFilter, SendEmptyMetadata>                                                                                 -7.50                               -7.50
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                          allocs_per_iteration    atm_add_per_iteration    atm_cas_per_iteration      cpu_time  locks_per_iteration      real_time  writes_per_iteration
-----------------------------------------------------------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_HpackEncoderEncodeHeader<SingleNonInternedBinaryElem<3>>/0/16k                                                                                -22.00                              -22.00
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                                allocs_per_iteration    atm_add_per_iteration    atm_cas_per_iteration      cpu_time  locks_per_iteration      real_time  writes_per_iteration
-----------------------------------------------------------------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_HpackEncoderEncodeHeader<EmptyBatch>/1/16k                                                                                                           +6.50                               +6.50
BM_HpackEncoderEncodeHeader<RepresentativeServerTrailingMetadata>/1/16k                                                                                +16.00                              +16.00
BM_HpackEncoderEncodeHeader<SingleInternedBinaryElem<1>>/0/16k                                                                                         +29.50                              +29.50
BM_HpackParserParseHeader<RepresentativeServerTrailingMetadata>                                                                                        +15.50                              +15.50
BM_IsolatedFilter<HttpClientFilter, NoOp>                                                                                                              +46.00                              +46.00
BM_IsolatedFilter<LoadReportingFilter, SendEmptyMetadata>                                                                                              +39.50                              +39.50
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                allocs_per_iteration    atm_add_per_iteration    atm_cas_per_iteration      cpu_time  locks_per_iteration      real_time  writes_per_iteration
---------------------------------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_IsolatedFilter<CompressFilter, NoOp>                                                                                 -6.50                               -6.50
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                        allocs_per_iteration    atm_add_per_iteration    atm_cas_per_iteration      cpu_time  locks_per_iteration      real_time  writes_per_iteration
---------------------------------------------------------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_HpackEncoderEncodeHeader<SingleInternedBinaryElem<31>>/0/16k                                                                                 -8.50                               -8.50
BM_HpackParserParseHeader<AddIndexedSingleStaticElem>                                                                                          +28.00                              +28.00
BM_SingleThreadPollOneFd_SpeedOfLight                                                                                                          -55.00                              -55.00
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                   allocs_per_iteration    atm_add_per_iteration    atm_cas_per_iteration      cpu_time  locks_per_iteration      real_time  writes_per_iteration
--------------------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_HpackEncoderInitDestroy                                                                               +150.50                             +150.50
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                 cpu_time    real_time
--------------------------------------  ----------  -----------
BM_ErrorCreateAndSetIntAndStr               -43.00       -43.00
BM_ErrorCreateAndSetStatus                  -32.00       -32.00
BM_ErrorCreateFromCopied                    -47.00       -47.00
BM_ErrorCreateFromStatic                    -28.50       -28.50
BM_MetadataRefUnrefAllocated                 -6.00        -6.00
BM_PollEmptyPollset_SpeedOfLight/1/100      -12.50       -12.50
BM_SliceIntern                              -20.00       -20.00
BM_SliceReIntern                             -9.50        -9.50
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                           cpu_time    real_time
--------------------------------  ----------  -----------
BM_ErrorGetStatusCode<ErrorNone>      +16.00       +16.00
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                                  cpu_time    real_time
-----------------------------------------------------------------------  ----------  -----------
BM_ErrorCreateAndSetStrLoop                                                   -6.00        -6.00
BM_HpackEncoderEncodeHeader<RepresentativeServerTrailingMetadata>/1/16k       +9.00        +9.00
BM_TransportStreamRecv/256k                                                -1112.00     -1112.00
BM_TransportStreamSend/256k                                                 -318.50      -318.50
```

ctiller commented:
Failed: #10659 #10643
